File: pglabs/load_vectors_in_db.js
// Import the 'pg' module for PostgreSQL interaction
const pg = require('pg');

// Configuration object for PostgreSQL connection
const config = {
  host: 'localhost',       // Database host
  database: 'postgres',    // Database name
  user: 'seti',        // Database user
  password: '',  //Replace this with the password of your PostgreSQL instance
  port: 5432               // Database port
}

// Create a PostgreSQL client with the config object
const client = new pg.Client(config);

// Array of sentences to be inserted into the database
let sentences = [
  "The horse is galloping",
  "The owl is hooting",
  "The rabbit is hopping",
  "The koala is munching",
  "The penguin is waddling",
  "The kangaroo is hopping",
  "The fox is prowling",
  "The parrot is squawking",
  "The turtle is crawling",
  "The cheetah is sprinting"
]

// Function to pad an array to a specified length with zeros
const padArray = (array) => {
    const length = 100;  // Desired length of the padded array

    // Check if the array is already the intended length or longer
    if (array.length >= length) {
        // If so, return a slice of the array to the intended length
        return array.slice(0, length);
    } else {
        // Otherwise, create a padding array filled with zeros to reach the intended length
        const padding = Array(length - array.length).fill(0);
        // Concatenate the original array with the padding array
        return array.concat(padding);
    }
}

// Function to convert a sentence into a numerical vector
const vectorize = (sentence) => {
    // Define a function to map characters to their ASCII values
    const charToInt = (char) => char.charCodeAt(0);

    // Convert the sentence into an array of ASCII values
    const vector = sentence.split('').map(charToInt);

    // Pad the resulting array to ensure it is 100 elements long
    return padArray(vector);
}

// Connect to the PostgreSQL database and create a table
client.connect()
  .then(async () => {
    console.log('Connected to PostgreSQL database');

    // SQL query to Create a table
    const createTableQuery = `
          CREATE TABLE IF NOT EXISTS mynewsentences (
            id SERIAL PRIMARY KEY,
            sentence TEXT,
            embedding INTEGER[]
          )`;

    // Execute the query to create the table
    await client.query(createTableQuery);
    console.log('Table created successfully');
    for (const sentence of sentences) {
        const embeddings = vectorize(sentence);
        await client.query('INSERT INTO mynewsentences (sentence, embedding) VALUES ($1, $2)', [sentence, embeddings])
      }

      const { rows } = await client.query(`SELECT * FROM mynewsentences`)

      for (let row of rows) {
          console.log(row.sentence);
      }
      process.exit(); // Exit the Node.js process
  })
  .catch(err => console.error('Error connecting to PostgreSQL database', err));

File: pglabs-python/load_vectors_in_db.py
import psycopg2
from psycopg2 import Error
import tensorflow_hub as hub
model = hub.load("https://tfhub.dev/google/universal-sentence-encoder/4")
print("Model loaded successfully")

# PostgreSQL database configuration
config = {
    'host': 'localhost',
    'database': 'postgres',
    'user': 'seti',
    'password': '', # Replace the password with the password of your postgres instance
    'port': 5432
}

# Sample sentences
sentences = ['Go to Try IBM watsonx.ai or Try watsonx.governance. If you sign up for watsonx.governance, you automatically provision watsonx.ai as well.',
                                  'Yes, when you sign up for IBM watsonx.ai, you automatically provision the free version of the underlying services: Watson Studio, Watson Machine Learning, and IBM Cloud Object Storage. When you sign up for IBM watsonx.governance, you automatically provision the free version of Watson OpenScale and the free versions of the services for IBM watsonx.ai.',
                                  'When you are ready to upgrade any of the underlying services for watsonx.ai or watsonx.governance, you can upgrade in place without losing any of your work or data. You must be the owner or administrator of the IBM Cloud account for a service to upgrade it. See Upgrading services on watsonx.'
]


def vectorize(sentence):
    return model([sentence])[0].numpy().tolist()

def add_data():
    try:
        # Connect to PostgreSQL database
        connection = psycopg2.connect(**config)

        # Create a cursor object
        cursor = connection.cursor()

        # drop old table if it exists 
        cursor.execute("DROP TABLE IF EXISTS mysentences;")

        # Define SQL query to create table
        create_table_query = '''
            CREATE TABLE IF NOT EXISTS mysentences (
                id SERIAL PRIMARY KEY,
                sentence TEXT,
                embedding FLOAT8[]
            );
        '''

        # Execute the query to create the table
        cursor.execute(create_table_query)
        connection.commit()
        print('Table created successfully')

        # Insert data into the table
        for sentence in sentences:
            embeddings = vectorize(sentence)
            insert_query = '''
                INSERT INTO mysentences (sentence, embedding)
                VALUES (%s, %s);
            '''
            cursor.execute(insert_query, (sentence, embeddings))
            connection.commit()

        # Fetch and display the data
        select_query = "SELECT * FROM mysentences;"
        cursor.execute(select_query)
        rows = cursor.fetchall()
        for row in rows:
            print(row[1], end='\n\n')  # Print the 'sentence' column

    except Error as e:
        print(f"Error connecting to PostgreSQL: {e}")

    finally:
        # Close the connection
        if connection:
            cursor.close()
            connection.close()
            print("PostgreSQL connection is closed")

if __name__ == "__main__":
    add_data()